feat(Core): Implement SP Bonus Coefficients from DBC

ZhengPeiRu21 opened the pull request:
<!-- First of all, THANK YOU for your contribution. -->

## Changes Proposed:
-  Implement loading SP Bonus Coefficients from DBC
-  Values that exist in the DB will override the DBC values. I have included SQL file to remove existing DB values that seem that have valid looking DBC values as those should be the most correct.
- Ported from TrinityCore: https://github.com/TrinityCore/TrinityCore/commit/1826437c094966ae339a154844df8587b2122888
- Co-authored by: ariel-

## Issues Addressed:
<!-- If your fix has a relating issue, link it below -->
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/12510

## SOURCE:
<!-- If you can, include a source that can strengthen your claim -->
See original issue

## Tests Performed:
<!-- Does it build without errors? Did you test in-game? What did you test? On which OS did you test? Describe any other tests performed -->
- Compiles
- Still needs testing


## How to Test the Changes:
<!-- Describe in a detailed step-by-step order how to test the changes -->
 
1. Use affected spells (see SQL file) and confirm that damage seems appropriate 

## Known Issues and TODO List:
<!-- Is there anything else left to do after this PR? -->

<!-- If you intend to contribute repeatedly to our project, it is a good idea to join our discord channel. We set ranks for our contributors and give them access to special resources or knowledge: https://discord.com/invite/DasJqPba)
     Do not remove the instructions below about testing, they will help users to test your PR -->
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

pklloveyou commented on src/server/game/Entities/Unit/Unit.cpp:
```diff
@@ -1753,7 +1753,7 @@ void Unit::DealMeleeDamage(CalcDamageInfo* damageInfo, bool durabilityLoss)
 
             if (Unit* caster = (*dmgShieldItr)->GetCaster())
-                damage = this->SpellDamageBonusTaken(caster, i_spellProto, damage, SPELL_DIRECT_DAMAGE);
+                damage = this->SpellDamageBonusTaken(caster, i_spellProto, damage, SPELL_DIRECT_DAMAGE, (*dmgShieldItr)->GetEffIndex());
```
This is wrong？

pklloveyou replied:
As I said below, this is obviously wrong

pklloveyou replied:
And there are many omissions that have not been modified

ZhengPeiRu21 replied:
Thank you, I now see that I incorrectly modified SpellDamageBonusTaken instead of adding the missing SpellDamageBonusDone. I will make the correction.

pklloveyou commented:
And "spell_dru_lifebloom"

pklloveyou commented:
Copying is useless. There is a difference between AC and TC. You need to check all "SpellDamageBonusDone AND SpellHealingBonusDone"

ZhengPeiRu21 commented:
> Copying is useless. There is a difference between AC and TC. You need to check all "SpellDamageBonusDone AND SpellHealingBonusDone"

Thank you, it seems I did not check all SpellHealingBonusDone. I think I did already correct all SpellDamageBonusDone.

ZhengPeiRu21 commented:
Thank you pklloveyou for the reviewing. 没有必要那么凶巴巴吧。^_^ I added the overlooked Druid spell and corrected DealMeleeDamage calculation. I reviewed all SpellDamageBonusDone and SpellHealingBonusDone calls and did not see any more that were missed. I think this now should please be ready for a re-review. Please let me know if you see any more change that needs to be make!

pklloveyou commented:
没有凶啊。可能是我英文不太好？？

pklloveyou commented:
-     if (GetEffect(EFFECT_1))
+    if (AuraEffect const* aurEff = GetEffect(EFFECT_1))

ZhengPeiRu21 commented:
> 没有凶啊。可能是我英文不太好？？

抱歉抱歉，应该是我自己英文不好，搞错了你的意思。当我没说过吧！

> 代码生成失败在这里 spell_druid.cpp
> 
> ```
> -    if (GetEffect(EFFECT_1))
> +    if (AuraEffect const* aurEff = GetEffect(EFFECT_1))
> ```

谢谢！现在这里已经很晚了但是我明天会看一眼，把他弄好。

acidmanifesto commented:
認為其他人會使用谷歌翻譯或不同的網站來關注這個對話？ 😂

heyitsbench commented:
> 認為其他人會使用谷歌翻譯或不同的網站來關注這個對話？ 😂

Nah, people wouldn't do that. 😛

Nyeriah commented:
> > 認為其他人會使用谷歌翻譯或不同的網站來關注這個對話？ 😂
> 
> Nah, people wouldn't do that. 😛

Yep. Np.

![image](https://user-images.githubusercontent.com/47818697/182008679-4838b003-9ab6-4854-8510-7acd217c5f48.png)

pklloveyou commented:
😂

ZhengPeiRu21 commented:
Updated to fix spell_dru_lifebloom.

acidmanifesto reviewed (approved):
謝謝，讓我們看看它是否全部變綠。

Maelthyrr commented:
Hi, PR tested:

- [X] Tested 50% spells and I see any issue from lvl 1 to 15

But I'm not sure how SP add by items is applied but it's seems that didn't work. I've setup a mage to lvl60 and add only 16916 (32 SP bonus), but with fireball rank2 I stay with default values. I get any dmg bonus.

ZhengPeiRu21 commented:
> 

Is anyone able to help with re-test this? The code should not change anything of how spell power bonus is applied; that is not related to the PR. This should only set the default value based on the coefficient from the DBC.

If spell power bonus is working without the PR changes but not working with the PR, that is of course a problem to fix! But I think the proper testing to test what this PR modifies is to change the coefficient values in spell_dbc and see if the spell power changes.

Gultask commented:
I haven't tested all spells, but the scaling seems ok.

Tested with Bloodmage Regalia giving 810 SP
Rank 1 spells have little effect from SP while higher rank spells have significantly more impact from SP.
